### PR TITLE
Filter/get method

### PR DESCRIPTION
**Pull request by quantranhong1999**

### CODE REVIEW

File: openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/model/FilterGet.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+case class FilterGetResponse(accountId: AccountId,
+                             list: List[Filter],
+                             notFound: FilterGetNotFound)
+
+case class FilterGetIds(value: List[String])
+
+case class Condition(field: Field, comparator: Comparator, value: String)
+
+case class Field(string: String)
```
Comment (chibenwa) [inline openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/model/FilterGet.scala:21]: extend AnyVal?

File: openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/model/FilterGet.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+case class Comparator(string: String)
```
Comment (chibenwa) [inline openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/model/FilterGet.scala:23]: extend AnyVal?

File: openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+class FilterGetMethod @Inject()(val metricFactory: MetricFactory,
+                                val sessionSupplier: SessionSupplier,
+                                val mailboxIdFactory: MailboxId.Factory,
+                                filteringManagement: FilteringManagement) extends MethodRequiringAccountId[FilterGetRequest] {
+
+  override val methodName: Invocation.MethodName = MethodName("Filter/get")
+  override val requiredCapabilities: Set[CapabilityIdentifier] = Set(LINAGORA_FILTER)
+
+  override def doProcess(capabilities: Set[CapabilityIdentifier], invocation: InvocationWithContext, mailboxSession: MailboxSession, request: FilterGetRequest): Publisher[InvocationWithContext] = {
+    getFilterGetResponse(request, mailboxSession).map(response => InvocationWithContext(
```
Comment (chibenwa) [inline openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala:59]: You don't need the `{}` brackets here?

File: openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      processingContext = invocation.processingContext))
+  }
+
+  override def getRequest(mailboxSession: MailboxSession, invocation: Invocation): Either[Exception, FilterGetRequest] =
+    FilterSerializer.deserializeFilterGetRequest(invocation.arguments.value) match {
+      case JsSuccess(filterGetRequest, _) => Right(filterGetRequest)
+      case errors: JsError => Left(new IllegalArgumentException(ResponseSerializer.serialize(errors).toString))
+    }
+
+  private def parseRuleFromJavaToScala(rule: JavaRule): Rule = {
```
Comment (chibenwa) [inline openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala:73]: You don't need the `{}` brackets here?

File: openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    Rule(Name(rule.getName),
+      Condition(Field(rule.getCondition.getField.asString()), Comparator(rule.getCondition.getComparator.asString()), rule.getCondition.getValue),
+      Action(AppendIn(parseMailboxIds(rule.getAction.getAppendInMailboxes.getMailboxIds))))
+  }
+
+  private def parseMailboxIds(mailboxIds: ImmutableList[String]) : List[MailboxId] = {
```
Comment (chibenwa) [inline openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala:79]: You don't need the `{}` brackets here?

File: openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    List(mailboxIdFactory.fromString(mailboxIds.get(0)))
+  }
+
+  private def retrieveFilters(username: Username) : SMono[Filter] = {
```
Comment (chibenwa) [inline openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala:83]: You don't need the `{}` brackets here?

File: openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+      .collectSeq()
+      .map(rules => Filter("singleton", rules.toList))
+  }
+
+  private def getFilterGetResponse(request: FilterGetRequest,
+                                  mailboxSession: MailboxSession): SMono[FilterGetResponse] =
+    request.ids match {
+      case None => retrieveFilters(mailboxSession.getUser).map(filter => FilterGetResponse(request.accountId, List(filter), FilterGetNotFound(List())))
+      case Some(ids) => if(ids.value.contains("singleton")) {
+        retrieveFilters(mailboxSession.getUser).map(filter => FilterGetResponse(request.accountId, List(filter), FilterGetNotFound(ids.value.filterNot(id => id.equals("singleton")))))
```
Comment (chibenwa) [inline openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala:95]: Can we add a line break before `.map` for readability

File: openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterGetMethodContract.scala [hunk 8]
```diff
@@ -0,0 +1,523 @@
...
+      .addUser(BOB.asString(), BOB_PASSWORD)
+
+    requestSpecification = baseRequestSpecBuilder(server)
+      .setAuth(authScheme(UserCredential(BOB, BOB_PASSWORD)))
+      .build()
+  }
+
+  @Test
+  def filterGetWithUserHaveExistingRulesShouldShowThoseRules(server: GuiceJamesServer): Unit = {
+
```
Comment (chibenwa) [inline openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterGetMethodContract.scala:39]: We do not need this line break?

File: openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterGetMethodContract.scala [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
+                     |        },
+                     |          "c1"]
+                     |    ]
+                     |}""".stripMargin
+
+    val response = `given`()
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when()
+      .post().prettyPeek()
```
Comment (chibenwa) [inline openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraFilterGetMethodContract.scala:67]: `.prettyPeek()` spotted please remoe them all...

**Review by chibenwa — COMMENTED** [review]
Looks good to me!

**Review by chibenwa — APPROVED** [review]

File: openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  override def getRequest(mailboxSession: MailboxSession, invocation: Invocation): Either[Exception, FilterGetRequest] =
+    FilterSerializer.deserializeFilterGetRequest(invocation.arguments.value) match {
+      case JsSuccess(filterGetRequest, _) => Right(filterGetRequest)
+      case errors: JsError => Left(new IllegalArgumentException(ResponseSerializer.serialize(errors).toString))
+    }
+
+  private def parseRuleFromJavaToScala(rule: JavaRule): Rule =
+    Rule(Name(rule.getName),
+      Condition(Field(rule.getCondition.getField.asString()), Comparator(rule.getCondition.getComparator.asString()), rule.getCondition.getValue),
+      Action(AppendIn(parseMailboxIds(rule.getAction.getAppendInMailboxes.getMailboxIds))))
```
Comment (LanKhuat) [inline openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala:76]: Can we define this method as part of the Rule class?

File: openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala [hunk 11]
```diff
@@ -70,21 +70,18 @@ class FilterGetMethod @Inject()(val metricFactory: MetricFactory,
...
-      Action(AppendIn(parseMailboxIds(rule.getAction.getAppendInMailboxes.getMailboxIds))))
-
   private def parseMailboxIds(mailboxIds: ImmutableList[String]) : List[MailboxId] = List(mailboxIdFactory.fromString(mailboxIds.get(0)))
 
   private def retrieveFilters(username: Username) : SMono[Filter] =
     SFlux.fromPublisher(filteringManagement.listRulesForUser(username))
-      .map(parseRuleFromJavaToScala)
+      .map(rule => Rule(Name(rule.getName),
+        Condition(Field(rule.getCondition.getField.asString()), Comparator(rule.getCondition.getComparator.asString()), rule.getCondition.getValue),
+        Action(AppendIn(parseMailboxIds(rule.getAction.getAppendInMailboxes.getMailboxIds)))))
```
Comment (LanKhuat) [inline openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala:79]: What i meant is move the parse method in Rule class so you can do something like: .map(Rule.fromjava(...)) :)

Comment (quantranhong1999) [reply openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala:79]: Actually if I move parseRuleFromJavaToScala method (to a object Rule to use static method like above, right?), i will need move parseMailboxsIds method also which use mailboxIdFactory(need to be inject). It's quite complicated

Comment (quantranhong1999) [reply openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala]: Please have a look again to see if the code is OK

Comment (LanKhuat) [reply openpaas-james/jmap/src/main/scala/com/linagora/openpaas/james/jmap/method/FilterGetMethod.scala]: i think fromJava() would be shorter and also more consistent with our current naming strategy.

**Review by LanKhuat — APPROVED** [review]

**Comment by Arsnael** [comment]
Not sure what commit of the submodule you point at... Please for now, do:
```
git submodule update
```
And then commit/push the change 
That should put you back to the actual commit on master for james-project

**Comment by quantranhong1999** [comment]
> Not sure what commit of the submodule you point at... Please for now, do:
> 
> ```
> git submodule update
> ```
> 
> And then commit/push the change
> That should put you back to the actual commit on master for james-project

I updated submodule to its last version in this commit : https://github.com/linagora/openpaas-james/pull/51/commits/e0b04bfca45ce5d70ce0cf16c4276ba2b4996704

This commit points to ec7b50b82c5386aec6ee115a3b4c1b3018f7b4a8 commit means : https://github.com/apache/james-project/commit/ec7b50b82c5386aec6ee115a3b4c1b3018f7b4a8

**Comment by Arsnael** [comment]
Update to the one used on master, not the latest one on apache/james-project. The reason being as long as the latest updates from the submodule are not tested, we can't rely on it (for example it's breaking the build atm)

**Comment by chibenwa** [comment]
Can you stash the fixups so that we can merge this?

**Comment by Arsnael** [comment]
This time the build is valid :)

Not sure why can't do an automatic merge though... Might need a stash and a bit of a cleanup

**Pull request merged**